[proposal] make the CLI available without cloning the repo

Eraz1997 opened the pull request:
hey 👋

This repo and the job you did is amazing! I wanted to contribute with this PR with something I'd value when it comes to some of my personal projects. I'm not sure this is something you want to upstream as well, so I marked it as a "proposal".

With all these changes, I actually just moved the `bin/train/main.py` script inside the package to make it available to users installing the `nam` PyPi package without having to clone the whole repo.

Since there's no "bin" inside the `bin` folder anymore, I moved the `bin/train` one at the root level.

Finally, I aligned the `README` with these changes (+ I removed some old, I guess, paragraphs)


I hope it can be of any help 💪

Eraz1997 commented:
Additional note, I’m honestly unsure whether running the CLI script outside of conda changes anything

sdatkinson commented:
I like this idea. However, your PR does unfortunately introduce a lot of issues that need to get addressed before I can merge it.

I'll need to find some time to review this.

Eraz1997 commented:
Oh, I’m sorry I haven’t noticed them! If you give me some pointers I can try to address the issues 💪

One thing I just noticed is that `pytorch-cuda` is missing if you don’t run the CLI inside the conda environment. I can think about a solution, if needed.

sdatkinson commented on nam/_version.py:
```diff
@@ -1 +1 @@
-__version__ = "0.7.4"
+__version__ = "0.8.0"
```
Pull latest

sdatkinson commented on .gitignore:
```diff
@@ -131,4 +131,4 @@ dmypy.json
 
 # Training outputs
-./bin/train/outputs/
+./train/outputs/
```
Don't move this stuff around. Let me handle it.

Eraz1997 replied:
I'll keep everything under the `bin` folder then - I thought of removing it since there was no more a "bin" there, strictly speaking, but sorry about that 👍

sdatkinson commented on tests/test_nam/test_train/test_cli.py:
```diff
@@ -173,22 +171,20 @@ def _setup_files(self, root_path: Path, device: _Device):
-                    str(self._output_path(tempdir, ensure=True)),
-                    "--no-show",
-                ]
+            run(
```
Revert; this doesn't check that the script works from the command line (and you broke it).

sdatkinson commented on nam/train/cli.py:
```diff
@@ -225,11 +225,15 @@ def main_inner(
 
 
-if __name__ == "__main__":
+def main():
```
[Nit] Now that there are 3 different "entry points" with different interfaces, it'd probably be good to docstring why they're all here and what they achieve [I can handle this.]

sdatkinson commented on train/easy_colab.ipynb:
[BLOCKING] Undo these moves

sdatkinson reviewed (changes requested):
See notes

Eraz1997 commented on README.md:
```diff
@@ -29,32 +29,9 @@ After installing the Python package, a GUI can be accessed by running `nam` in t
 
 ### The command line trainer (all features)
 
-Alternatively, you can clone this repo to your computer and use it locally.
```
Sounds good 👌 My obviously questionable suggestion is to clarify what is a `CONTRIBUTING` guide and what are usage guidelines. This paragraph looks more like user guidelines, so I'd specify that.

I see you kinda split those two parts in the new docs [here](https://neural-amp-modeler.readthedocs.io/en/latest/tutorials/command-line.html).

I'll leave them as they are and possibly add just a paragraph 💪

Eraz1997 commented on README.md:
```diff
@@ -100,28 +77,43 @@ is being trained. The example used here uses a `feather` configured `wavenet`.
-under the hood as a modeling framework, and you can control many of the Pytorch Lightning configuration options from `bin/train/inputs/learning/default.json`
+under the hood as a modeling framework, and you can control many of the Pytorch Lightning configuration options from `train/inputs/learning/default.json`
 
-#### Export a model (to use with [the plugin](https://github.com/sdatkinson/NeuralAmpModelerPlugin))
```
Yep, sorry about that! 🙏

Eraz1997 commented:
I rewrote the PR from scratch as I thought it'd have been simpler - and indeed I came up with way fewer and leaner changes IMHO 🚀 

Sorry for the first round of review, it must have been tough as the PR was a bit messy and changed a lot of things I didn't quite check you agreed with before 🙏 

What mainly changes from the "previous version of the PR" is:

- I kept files at their place
- I didn't remove any documentation, just added a tiny paragraph about pip and the new `nam-cli` command

> I kept files at their place

I guess this is what you were referring to when you mentioned the broken test - indeed `nam-cli` was working perfectly fine, but if you meant the `python bin/train/main.py ...` command, then yes, it broke because I intentionally replaced it with `nam-cli`, but indeed I should have checked it with you first! 🙏 



Thanks!! 😊

sdatkinson reviewed (commented):
Overall, this looks great!

I can take care of the nits and checks myself, but thanks for taking the initiative on this--I like it 👍🏻

Eraz1997 commented on bin/train/main.py:
```diff
@@ -2,234 +2,8 @@
-        plot(model, dataset_validation, show=not no_show)
-    # Export!
-    model.net.export(outdir)
+from nam.train.cli import run
```
Sounds great to me 🚀

Eraz1997 commented:
Great!! LMK if you want me to help somehow 🚀 Thank you!

sdatkinson commented:
@Eraz1997 on second thought, if you can grab those changes before Thursday you'll probably get to it before me :)

Eraz1997 commented:
@sdatkinson unfortunately I don't have band this week, LMK if it can be of any help next week as well!

sdatkinson commented:
Got the nits and did a bit of cleanup on the documentation in [Eraz1997/main](https://github.com/sdatkinson/neural-amp-modeler/tree/Eraz1997/main).

Closing this PR and we'll merge there 👍🏻

Eraz1997 commented:
Perfect!! 🚀
